### DISCUSSION TITLE
Problems while executing read/write queries against distributed databases(master/slaves)

### DISCUSSION THREAD

**Discussion by wiseaidev** [sqlalchemy/sqlalchemy#8713] [Usage Questions] [answered]
Hey everyone. I am currently in the process of [migrating from `databases` to SQLAlchemy](https://github.com/wiseaidev/fastapi-singlestore-backend/pull/32). However, I am running into issues when executing an `INSERT` statement after a `SELECT` statement. The following is the error message:

<details>
<summary><code>error message</code></summary>

```sh
qlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2549, 'Writing to a reference table in multi-statement transaction is not permitted on child aggregators. Try the command again on the master aggregator.')
[SQL: 
        INSERT INTO
            access_tokens (
                user,
                token,
                creation_date,
                token_status
            )
        VALUES
            (
                %s,
                %s,
                %s,
                1
            )
    ]
[parameters: (2251799813685268, 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJjY0BjYy5jYyIsImV4cCI6MTY2Njc0MjA3Nn0.yBzEQvqDY9L8OekE-q6IvKhb3D98rlSlhCSCD93BjhY', datetime.datetime(2022, 10, 25, 17, 54, 36, 443329))]
```

</details>

This exception is being thrown after calling [the execute function](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/9da092dfa358582f22960e77813fde71332add26/app/auth/crud.py#L130) on a session instance:

<details>
<summary><code>insert query</code></summary>

```python
    query = """
        INSERT INTO
            access_tokens (
                user,
                token,
                creation_date,
                token_status
            )
        VALUES
            (
                :user,
                :token,
                :creation_date,
                1
            )
    """
    values = {
        "user": user_obj.id,
        "token": access_token["access_token"],
        "creation_date": datetime.datetime.utcnow(),
    }
    await session.execute(text(query), values)
```

</details>

Note that sessions are generated per request, by [dependency injection](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/9da092dfa358582f22960e77813fde71332add26/app/auth/router.py#L49), using a [`session_factory`](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/9da092dfa358582f22960e77813fde71332add26/app/utils/session.py#L66) that [creates](https://github.com/wiseaidev/fastapi-singlestore-backend/blob/9da092dfa358582f22960e77813fde71332add26/app/utils/dependencies.py#L27) a session for each request made.

The error message suggests running the insert command on a master aggregator. However, I am not sure how to do so.

On the positive side, I was successfully able to migrate the `register` endpoint to use SQLAlchemy for read/write data. It would be great if I could migrate the whole project soon.

That's pretty much it. I hope that I have provided enough info regarding this issue. If it is not the case, lemme know. Thanks for taking this discussion into consideration.

**Comment by zzzeek** [comment, marked as answer]
hi there -

this error message appears to be specific to the [Singlestore database](https://support.singlestore.com/hc/en-us/articles/360060908252-SELECT-that-reads-from-a-reference-table-on-a-leaf-inside-a-multi-statement-transaction).  You'd want to contact that vendor and get help from them, this is nothing that SQLAlchemy has control over.

**Reply by wiseaidev** [reply]
I am going to open a ticket there. But, what boggles my mind is how it was working with `databases`. Do you know how to apply the third point in SQLAlchemy:

```sh
3. Sometimes, it is possible to run the statements either as separate transactions or outside a multi-statement transaction.
```

Does it mean that I need to run the insert statement in a separate session? However, I tried doing so, but it didn't work.

**Reply by zzzeek** [reply]
it depends on if this mode is triggered by using autocommit mode.

take a look at https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#setting-isolation-for-a-sessionmaker-engine-wide.   in the second example:

```py
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

eng = create_engine("postgresql://scott:tiger@localhost/test")

autocommit_engine = eng.execution_options(isolation_level="AUTOCOMMIT")

transactional_session = sessionmaker(eng)
autocommit_session = sessionmaker(autocommit_engine)
```

that makes two sessionmakers.  the "autocommit_session" , when you run statements with Session objects made from that factory, will run without any transaction at all at the driver level.   

since "databases" wrote their own dialects, they might have been using this autocommit mode, or doing something else unusual.

**Reply by wiseaidev** [reply]
Ayo! the `isolation_level="AUTOCOMMIT"` option works. less gooo!
